### ISSUE TITLE
Associations named 'type' break other columns

### ISSUE THREAD

**Issue by bradshja** [balderdashy/waterline#788]
Creating an association named 'type' (maybe only belongsTo associations) causes other columns to misbehave by including the association's data.  As an example, I have models like this:

Account.js:
```
module.exports = {
    ...
    attributes: {
        account_id: {
            type: 'integer',
            autoIncrement: true,
            primaryKey: true
        },
        type: {
            model: 'accounttype',
            columnName: 'type_id'
        },
       devices: {
            model: 'device',
            via: 'account'
       }
    }
};
```
AccountType.js:
```
module.exports = {
    ...
    attributes: {
        type_id: {
            type: 'integer',
            autoIncrement: true,
            primaryKey: true
        },
        type: {
           type: 'string'
        }
    }
};
```
Device.js:
```
module.exports = {
    ...
    attributes: {
        device_id: {
            type: 'integer',
            autoIncrement: true,
            primaryKey: true
        },
        account: {
            model: 'account',
            columnName: 'account_id'
        },
       mac_address: {
            type: 'string'
       }
    }
};
```

Given those models, if I retrieve an account with the API, I expect to get something like this:
```
{
  "devices": [ 
     {
      "device_id": 561665222302,
      "mac_address": "00:02:02:4f:b1:59",
      "account_id": 105149
    }
  ],
  "type": {
    "type_id": 8,
    "type": "Employee"
  },
  "status": {
    "status_id": 1,
    "status": "Active"
  },
  "account_id": 105149,
  "display_name": "Jordan Bradshaw"
}
```
but I get this instead:
```
{
  "devices": [],
  "type": {
    "type_id": 8,
    "type": "Employee"
  },
  "status": {
    "status_id": 1,
    "status": "Active"
  },
  "account_id": [
    {
      "device_id": 561665222302,
      "mac_address": "00:02:02:4f:b1:59",
      "account_id": 105149
    }
  ],
  "display_name": "Jordan Bradshaw"
}
```
As you can see, the devices association is populating the account_id property, which is supposed to just be a numerical column as the primary key of the accounts collection.

After much investigation, I've discovered that this is caused by the Transformation.serialize function being applied to the collection's schema in waterline/core/index.js on line 104.  The serialization function sees the type column and recursively replaces all instances of 'type' with 'type_id.'  I'm not sure how this causes the wackiness above, but removing the serialization line fixes the problem, at the possible expense of causing problems elsewhere that I haven't seen.

It doesn't look like Transformation.serialize can be made non recursive since it's used in other places, so some other function might should be used to transform the collection's schema.  If I had to guess, this probably prevents any association that uses reserved field names from working right.  'type', 'primaryKey', 'model', 'collection' and so on probably don't work either.

**Comment by dasher** [comment]
type as column name would be a reserved word - change to something else
like ofType and your problem will go away without needing to change the
core code.

On 8 January 2015 at 23:46, bradshja <notifications@github.com> wrote:

> Creating an association named 'type' (maybe only belongsTo associations)
> causes other columns to misbehave by including the association's data. As
> an example, I have models like this:
>
> Account.js:
>
> module.exports = {
>     ...
>     attributes: {
>         account_id: {
>             type: 'integer',
>             autoIncrement: true,
>             primaryKey: true
>         },
>         type: {
>             model: 'accounttype',
>             columnName: 'type_id'
>         },
>        devices: {
>             model: 'device',
>             via: 'account'
>        }
>     }
> };
>
> AccountType.js:
>
> module.exports = {
>     ...
>     attributes: {
>         type_id: {
>             type: 'integer',
>             autoIncrement: true,
>             primaryKey: true
>         },
>         type: {
>            type: 'string'
>         }
>     }
> };
>
> Device.js:
>
> module.exports = {
>     ...
>     attributes: {
>         device_id: {
>             type: 'integer',
>             autoIncrement: true,
>             primaryKey: true
>         },
>         account: {
>             model: 'account',
>             columnName: 'account_id'
>         },
>        mac_address: {
>             type: 'string'
>        }
>     }
> };
>
> Given those models, if I retrieve an account with the API, I expect to get
> something like this:
>
> {
>   "devices": [
>      {
>       "device_id": 561665222302,
>       "mac_address": "00:02:02:4f:b1:59",
>       "account_id": 105149
>     }
>   ],
>   "type": {
>     "type_id": 8,
>     "type": "Employee"
>   },
>   "status": {
>     "status_id": 1,
>     "status": "Active"
>   },
>   "account_id": 105149,
>   "display_name": "Jordan Bradshaw"
> }
>
> but I get this instead:
>
> {
>   "devices": [],
>   "type": {
>     "type_id": 8,
>     "type": "Employee"
>   },
>   "status": {
>     "status_id": 1,
>     "status": "Active"
>   },
>   "account_id": [
>     {
>       "device_id": 561665222302,
>       "mac_address": "00:02:02:4f:b1:59",
>       "account_id": 105149
>     }
>   ],
>   "display_name": "Jordan Bradshaw"
> }
>
> As you can see, the devices association is populating the account_id
> property, which is supposed to just be a numerical column as the primary
> key of the accounts collection.
>
> After much investigation, I've discovered that this is caused by the
> Transformation.serialize function being applied to the collection's schema
> in waterline/core/index.js on line 104. The serialization function sees the
> type column and recursively replaces all instances of 'type' with
> 'type_id.' I'm not sure how this causes the wackiness above, but removing
> the serialization line fixes the problem, at the possible expense of
> causing problems elsewhere that I haven't seen.
>
> It doesn't look like Transformation.serialize can be made non recursive
> since it's used in other places, so some other function might should be
> used to transform the collection's schema. If I had to guess, this probably
> prevents any association that uses reserved field names from working right.
> 'type', 'primaryKey', 'model', 'collection' and so on probably don't work
> either.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/waterline/issues/788>.
>



-- 
Regards,

David Ashwood
david@CrucialDivide.com
+34 608 929 043

**Comment by bradshja** [comment]
Sure, that would work, but we're replacing an existing, in-house developed API that has a pretty substantial amount of client code that expects the association to be named 'type'.  I can understand that there are going to be column names that aren't acceptable, but type (and model and collection even, if they don't work either), aren't that exotic and it would be really nice if they worked.  Especially since regular columns named 'type' seem to work just fine.

Is there no chance this can be fixed in the core code?

**Comment by dasher** [comment]
The current ones - like "type" you have found should be documented - just
to stop the pain you experienced tracking this one down.
The dev team can then decide if they should be prefixed or changed to stop
collisions later.
With some storage engines like mongo - it can be tricky as they allow
nested documents, and as we all know, naming stuff can be hard - and with
mongo you can find yourself struggling to find meaningful names.

A prefix would work help avoid collisions - but can be painful when
migrating datasets between the version boundary where the change happens.

If you are replacing or migrating an existing system - and you control all
the IO to the back end (meaning they'll be using your interface to the db)
- you can translate the column name at your own boundary.  Then it doesn't
matter what the col name is - they'll use type and you'll convert it as
needed to whichever name you decide to use.



On 9 January 2015 at 00:39, bradshja <notifications@github.com> wrote:

> Sure, that would work, but we're replacing an existing, in-house developed
> API that has a pretty substantial amount of client code that expects the
> association to be named 'type'. I can understand that there are going to be
> column names that aren't acceptable, but type (and model and collection
> even, if they don't work either), aren't that exotic and it would be really
> nice if they worked. Especially since regular columns named 'type' seem to
> work just fine.
>
> Is there no chance this can be fixed in the core code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/waterline/issues/788#issuecomment-69270539>
> .
>



-- 
Regards,

David Ashwood
david@CrucialDivide.com
+34 608 929 043

**Comment by bradshja** [comment]
Transforming the column name in code is something I considered and if this can't or won't be fixed in the core code then that's probably what we'll have to end up doing.  Better than trying to maintain a local patch anyway.

Definitely agreed about documenting it though.  Searching to see if it was a reserved column name was one of the first things I did when I realized which association was causing the trouble.  I did find an array of reserved words in some file deep in Waterline's code, but it wasn't clear that's what it was supposed to indicate.

**Comment by particlebanana** [comment]
This should be fixed. If it's a valid javascript key then you should be allowed to use it. It might take a bit of refactoring but shouldn't cause any breaking changes. @bradshja if you have something that works and doesn't completely remove the transformations would you mind submitting a pr?

**Comment by bradshja** [comment]
I just submitted pull request #789 that fixes my case at least.  I ended up adding some small changes to Transformation.serialize that lets you restrict the depth that it serializes to (no max by default).  By limiting the depth of serialization to 1 level in lib/core/index.js, the types of other columns are no longer clobbered by the 'type' association's columnName.

It's untested in general, but it probably fixes other related cases too.
